### ISSUE TITLE
Run Ngninx Kubernetes Ingress controller with `readOnlyRootFilesystem: true` 

### ISSUE THREAD

**Issue by abhisek-dwivedi** [nginxinc/kubernetes-ingress#1677]
**Is your feature request related to a problem? Please describe.**

Users should be able to run nginx Kubernetes ingress controller with minimum permissions in the security context. Current configuration runs with `AllowPrivilegeEscalation: true` and `readOnlyRootFilesystem: false` but there should be a way to restrict the permissions to a bare minimum.

**Describe the solution you'd like**

I was able to remove `AllowPrivilegeEscalation: true` by changing the hard code privileged ports (80, 443) in code & templates and building a custom docker image out of it. But for the `readOnlyRootFilesystem: false` workaround I'm trying to change `/etc/nginx` directory where all configurations are stored/created to one emptyDir volume.  

But after these changes, I'm facing sock connection failure and other issues in IC pod. 
Eg: 
`F0616 14:26:09.191702       1 manager.go:284] Could not get newest config version: could not get expected version: 0 after 4s`

Is it possible to run IC with `readOnlyRootFilesystem: true` ? or some specific steps are required to do so ?

Aha! Link: https://nginx.aha.io/features/IC-97

**Comment by pleshakov** [comment]
Hi @abhisek-dwivedi 

> Is it possible to run IC with readOnlyRootFilesystem: true ? or some specific steps are required to do so ?

We haven't tested if it's possible to run an IC pod with `readOnlyRootFilesystem: true`, so we don't have any specific steps or documentation.

Note that in addition to `/etc/nginx` folder, the following folders are also used for writes:
- `/var/cache/nginx` - NGINX stores cache related files if caching is enabled
- `/var/lib/nginx` - NGINX stores various temp files here. Also, the IC configures NGINX to open a number of auxiliary unix socket in that folder.

**Comment by abhisek-dwivedi** [comment]
Okay, thanks. 

Would it possible to merge of all these folders to just one common folder by setting some variable in code and mounting that path as an emptyDir volume externally? 
If not, how about mounting 3 emptyDir volumes for each path wherever IC pod writes.

Not sure if that's not going to work. Let me know your thoughts.

Thanks

**Comment by brianehlert** [comment]
Hello @abhisek-dwivedi 
I understand the readOnlyRootFileSystem ask.
I am thinking more about what you are trying to achieve.
Would support for the nginx user directive or running nginx as non-root meet the desired outcome?
Or, is the primary issue around "AllowPrivilegeEscalation"

I am really trying to fully understand what you are trying to accomplish as opposed to focusing on one solution that you are asking about.  As there are other possible approaches.

**Comment by abhisek-dwivedi** [comment]
Yeah totally agree that there are other possible good approaches. I was just playing around with IC to make it work as per my use case. 

I have a product that runs along with IC, so you can say there is a hard dependency on it.  But now the problem is that my product runs with minimum Pod Security permission without `AllowPrivilegeEscalation` and `readOnlyRootFileSystem`. But now there are some customer's k8s environments that are tightly constrained using PSP or any other Policy engine that doesn't allow the above permission. So in that case IC controller breaks or not allowed to run. As a result, my product functionality breaks.

Also, normal helm chart test/linter tools flag these kinds of permission as well.

 FYI: I am able to get pass-through root user and  `AllowPrivilegeEscalation` permission by making some changes in the code and docker file. Now my IC runs with a non-root user and `AllowPrivilegeEscalation: false` but I am stuck on removing `readOnlyRootFileSystem` permission

**Comment by pleshakov** [comment]
Note about `AllowPrivilegeEscalation`

To allow NGINX to bind to privileged ports like 80 and 443,  we add the `NET_BIND_SERVICE` capability to the nginx binary (https://github.com/nginxinc/kubernetes-ingress/blob/master/build/Dockerfile#L267) and also allow this capability in the pod security context -- https://github.com/nginxinc/kubernetes-ingress/blob/master/deployments/deployment/nginx-ingress.yaml#L42-L46  When the IC container starts, the container runtime launches the IC process, which in turn starts NGINX. Because permitted capabilities of NGINX are different from permitted capabilities of IC (the IC process is not allowed to bind to privileged ports), it is necessary to configure `allowPrivilegeEscalation: true`. 

IC  process (`cat /proc/1/status | grep Cap` in the container):
```
CapInh: 0000000000000400
CapPrm: 0000000000000000
CapEff: 0000000000000000
CapBnd: 0000000000000400
CapAmb: 0000000000000000
```

NGINX master process (`cat /proc/<pid-of-NGINX-master>/status | grep Cap` in the container):
```
CapInh: 0000000000000400
CapPrm: 0000000000000400
CapEff: 0000000000000400
CapBnd: 0000000000000400
CapAmb: 0000000000000000
```

**Comment by sigv** [comment]
I am proposing a PR for the title of this feature request (readonly rootfs).

However there is a second item in the feature request bundled in, `allowPrivilegeEscalation: true`. As [previously explained](https://github.com/nginxinc/kubernetes-ingress/issues/1677#issuecomment-866499090), this is done for Nginx to bind ports 80 and 443. However, [Docker v20.10.0](https://docs.docker.com/engine/release-notes/20.10/#20100) (released 2020-12-08) should support binding privileged ports with no capabilities due to moby/moby#41030 merged as moby/moby@888da28d42f7d0f9fa250dd8a75d51c2a6cf3098 which sets `sysctl net.ipv4.ip_unprivileged_port_start=0`.
Should I spin off a second tracking issue for evaluation of removing `NET_BIND_SERVICE` and `allowPrivilegeEscalation` from Nginx IC?
